[auton] Navigation New Architecture & Refactoring

qhdwight opened the pull request:
Following: https://github.com/mmartin4972/mrover-workspace/blob/navDesign/NewNavDesign.md
- Use smart pointers instead of raw
- Fix some minor UB with relation to field instantiation
- Separate external rover state into Environment class instead of holding it directly inside the state machine
- ^^ Same as above also with course, in CourseProgress

qhdwight commented:
Right now it is close to compiling but I would like some feedback on resolving the last couple problems @niwhsa9

mmartin4972 commented on jetson/nav/utilities.cpp:
```diff
@@ -3,165 +3,135 @@
 // Converts the input radians to degrees.
-double radianToDegree( const double radian )
-{
+double radianToDegree(const double radian) {
```
You removed the const for this in the header file

qhdwight replied:
https://stackoverflow.com/questions/51353648/const-qualification-of-parameters-in-function-declaration

niwhsa9 replied:
seems like you'd want it in the header since thats the interface specification for other programmers though...

mmartin4972 commented on jetson/nav/utilities.cpp:
```diff
@@ -3,165 +3,135 @@
-        return ( mod + modulus );
-    }
-    return mod;
+double mod(const double degree, const int modulus) {
```
you removed the const for this in the header file

qhdwight replied:
https://stackoverflow.com/questions/51353648/const-qualification-of-parameters-in-function-declaration

mmartin4972 commented on jetson/nav/utilities.cpp:
```diff
@@ -3,165 +3,135 @@
-double degreeToRadian( const double degree, const double minute )
-{
-    return ( PI / 180 ) * ( degree + minute / 60 );
+double degreeToRadian(const double degree, const double minute) {
```
you removed the const for this in the header file

mmartin4972 reviewed (commented):
The changes you made look good.

I'm concerned however that you've made all these changes but don't have the code compiling. Because we don't want to mess anything up like Carmen said you need to break these changes apart, apply them individually, then run the unit tests on them. The reason we wanted to make the testing script first was so that we can test all of these changes incrementally. Please break it apart into 5 changes and unit test each individually:
1. moving smart pointers to pointers
2. seems like you ran some auto formatting tool, so have a step for that
3. looked like you refactored a few functions, so make the manual refactoring of these its own step
4. add the environment class modifying only the necessary LCM callbacks
5. add the course class and remove the old LCM callbacks

I'd like to see you make each of these stages of changes then test them. It will be easier for us to review if you break your commits down into this way. It will also make us more confident that no logic was changed since we did everything incrementally.

qhdwight commented:
Will address above for for viewing changes use the whitespace-cognizant mode on GitHub: https://github.com/umrover/mrover-workspace/pull/972/files?diff=unified&w=1

It is much easier to tell what changed semantically

niwhsa9 commented:
I agree with Matt about breaking this up into a couple PRs. I will hold off on my review until then

qhdwight commented:
Removed logic of maintaining an old and new rover state, removed the actual rover status class and just moved it into rover, decoupled further by extracting rover out of state machine

Still some funky behavior when it reaches it's first waypoint we have to iron out.

qhdwight commented:
Merged with newest auton-integration

zwgold reviewed (approved):
works well on sim and on the rover

cgiger00 commented on jetson/nav/CMakeLists.txt:
```diff
@@ -0,0 +1,28 @@
+cmake_minimum_required(VERSION 3.8)
```
can jarvis still build nav?

qhdwight replied:
Yes it builds on both. I can remove cmake if you'd like to keep things clean though

cgiger00 replied:
i think it's fine to keep in, can you add a comment in the readme or smth explaining why both are there/when one would the cmake file

cgiger00 commented on jetson/nav/layer2_test.cpp:
```diff
@@ -1,24 +1,24 @@
-
-    return 0;   //std::cout << layer2.state << std::endl;
-}
+////layer2 Test
```
can this be in a .gitignore?

qhdwight replied:
it should be deleted all together actually

cgiger00 commented on ansible/roles/jetson_dev/tasks/main.yml:
```diff
@@ -22,4 +22,5 @@
       - libgirepository1.0-dev
       - gfortran
       - xdotool
+      - libeigen3-dev
```
why is this in this PR?

cgiger00 commented on jetson/nav/rover.hpp:
```diff
@@ -11,15 +11,17 @@
 #include "pid.hpp"
 
-using namespace rover_msgs;
 using namespace std;
```
can you get rid of this & just add the specific members of std that you need? like using std::vector? etc

qhdwight replied:
sure

qhdwight replied:
done

cgiger00 commented on jetson/nav/rover.hpp:
```diff
@@ -63,178 +65,118 @@ enum class NavState
+
+    // The rover's current target information from computer
+    // vision.
+    Target mTargetLeft{-1.0, 0.0, 0};
```
is -1.0, 0.0, 0 the default? can this just be the default initialization for Target?

qhdwight replied:
It's an lcm auto generated, is that possible?

cgiger00 replied:
wym lcm auto generated?

qhdwight replied:
Target.hpp is auto generated by LCM, so I can't add a ctor I don't think that would set  it to -1.0, 0.0, 0

cgiger00 commented on pip_deps/jetson_requirements.txt:
```diff
@@ -1,2 +1,2 @@
 numpy==1.16.2
-odrive==0.5.4
+odrive==0.5.1
```
interesting that this change is in this PR?

cgiger00 commented on jetson/nav/.idea/.gitignore:
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
wait what are the .idea files?

qhdwight replied:
Oops, removed, must've been snuck in by CLion

cgiger00 reviewed (commented):
resolsved

cgiger00 reviewed (approved):
ok lgtm to merge
